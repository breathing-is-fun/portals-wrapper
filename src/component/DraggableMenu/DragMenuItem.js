import React, { Component } from 'react';

import { Menu } from 'antd';

export default class DragMenuItem extends Component {
    handleDragStart = (event, item) => {
    	event.dataTransfer.setData('menuItemToGrid', JSON.stringify(item));
    }

    render = () => {
    	const { item } = this.props;
    	const { draggable = true, text, id, group } = item;
    	const propsParam = {
    		key: group + id,
    		draggable,
    		onDragStart: e => this.handleDragStart(e, item),
    	};
    	const itemProps = Object.assign({}, this.props, propsParam);

    	return <Menu.Item { ...itemProps }>{ text }</Menu.Item>;
    }
}