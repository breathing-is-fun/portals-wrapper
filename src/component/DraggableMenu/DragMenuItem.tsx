import React, { Component } from 'react';
import { Menu } from 'antd';

export interface DragMenuItemProps {
  item?: any;
}
export default class DragMenuItem extends Component<DragMenuItemProps> {
  handleDragStart = (event: any, item: any) => {
    event.dataTransfer.setData('menuItemToGrid', JSON.stringify(item));
  };

  render = () => {
    const { item } = this.props;
    const { draggable = true, text, id, group } = item;
    const propsParam = {
      key: group + id,
      draggable,
      onDragStart: (e: any) => this.handleDragStart(e, item),
    };
    const itemProps = Object.assign({}, this.props, propsParam);

    return <Menu.Item {...itemProps}>{text}</Menu.Item>;
  };
}
