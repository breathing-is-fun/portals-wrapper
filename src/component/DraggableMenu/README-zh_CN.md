[English](./README.md) | 简体中文

# 简介

- 就像名字里写的一样，这是一个可拖拽的菜单

- 也可以像普通的Menu一样用，只要数据结构对，它可以帮你剩下排菜单结构的时间

- 基于 [antd的Menu](https://ant.design/components/menu-cn/)
  
# API  

| 名称 | 说明 | 类型 | 默认值 |
| --- | --- | --- | --- |
| selectedKeys | 当前选中的菜单项 key 数组 | [] | - |
| menuDatas | 菜单数据，对象格式见[下方](#menuDatas中的对象) | [] | - |
| openKeys | 当前展开的 SubMenu 菜单项 key 数组 | [] | - |
| onClick | 菜单项点击事件，具体可以看看 [这里](../../modules/ModuleEdit/index.js) 的 `handleMenuClick` | function (group: string, selectedKeys: string[], id: string) | - |
| onOpenChange | SubMenu 展开/关闭的回调 | function(openKeys: string[]) | - |
| shellStyleDatas | 用于控制拖拽结束时 [`Shell`](../Shell/index.js) 的样式，尚未实装 | [] | - |
| onSave | [点击左上保存时的事件](../../modules/ComponentEdit/index.js)，多用于保存右侧布局数据 | function () | - |
| type | 值为 `component` 时左上显示保存，否则跳转到 [`Display`](../../modules/Display/index.js) | string: oneOf(['component', 'module']) | - |
  
## menuDatas中的对象

| 名称 | 说明 | 类型 | 默认值 |
| --- | --- | --- | --- |
| group | 组别，用于分组 | string | - |
| groupname | 组别中文，分组后菜单项显示名称 | string | - |
| id | 菜单项唯一标识 | string or number | - |
| key | 用途暂时同id，必需 | - | - |
| order | 菜单项顺序，尚未实装 | string or number | - |
| text | 用途暂时同groupname，必需 | - | - |

# 示例

![img](../../../docs/img/DraggableMenu.gif)