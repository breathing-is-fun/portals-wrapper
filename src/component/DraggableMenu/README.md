English | [简体中文](./README-zh_CN.md)

# Introduction

- Just like its name, this is a draggbale menu.

- You can also use it as normal Menu. As long as data structure is correct, it can help you transform data structure to save your time

- Base on [antd的Menu](https://ant.design/components/menu/)
  
# API  

| name | description | type | default value |
| --- | --- | --- | --- |
| selectedKeys | array with the keys of currently selected menu items | [] | - |
| menuDatas | menu datas，the format of object in array is [here](#menuItemData) | [] | - |
| openKeys | array with the keys of currently opened sub menus | [] | - |
| onClick | click event of menu item，[there for detail](../../modules/ModuleEdit/index.js) about `handleMenuClick` | function (group: string, selectedKeys: string[], id: string) | - |
| onOpenChange | called when open/close sub menu | function(openKeys: string[]) | - |
| shellStyleDatas | 用于控制拖拽结束时 [`Shell`](../Shell/index.js) 的样式，尚未实装 | [] | - |
| onSave | [点击左上保存时的事件](../../modules/ComponentEdit/index.js)，多用于保存右侧布局数据 | function () | - |
| type | 控制左上显示保存或跳转到 [`Display`](../../modules/Display/index.js) | string: oneOf(['component', 'module']) | - |
  
## menuItemData

| 名称 | 说明 | 类型 | 默认值 |
| --- | --- | --- | --- |
| group | 组别，用于分组 | string | - |
| groupname | 组别中文，分组后菜单项显示名称 | string | - |
| id | 菜单项唯一标识 | string or number | - |
| key | 用途暂时同id，必需 | - | - |
| order | 菜单项顺序，尚未实装 | string or number | - |
| text | 用途暂时同group，必需 | - | - |

# 示例

![img](../../../docs/img/DraggableMenu.gif)