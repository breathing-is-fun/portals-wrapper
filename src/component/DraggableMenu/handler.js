/*
 * @Author: zy9@github.com/zy410419243
 * @Date: 2018-10-25 14:29:08
 * @Last Modified by: zy9
 * @Last Modified time: 2018-10-30 11:10:23
 */
const handleMenuGroup = dataSource => {
	let result = [], groupDatas = [], tempGroupDatas = [];

	// 分出有哪几组及组别名称
	for(let item of dataSource) {
		const { group, groupname: groupName } = item;

		tempGroupDatas.push(JSON.stringify({ group, groupName }));
	}
	tempGroupDatas = Array.from(new Set(tempGroupDatas));

	for(let item of tempGroupDatas) {
		groupDatas.push(JSON.parse(item));
	}

	// 凑菜单数据结构，[{ groupName: '', children: [{ text: '', url: '' }] }]
	for(let item of groupDatas) {
		const { group, groupName } = item;
		let childResult = { groupName, group, children: [] };

		for(let jtem of dataSource) {
			const { group: groupChild } = jtem;
			const { children } = childResult;

			if(groupChild == group) {
				children.push(jtem);
			}
		}

		// if(childResult.children && childResult.children.length == 1) {
		// 	delete childResult.children;
		// }

		result.push(childResult);
	}

	return result;
};

export { handleMenuGroup };