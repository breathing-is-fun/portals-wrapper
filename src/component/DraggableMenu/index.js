/*
 * @Author: zy9@github.com/zy410419243
 * @Date: 2018-09-28 17:29:59
 * @Last Modified by: zy9
 * @Last Modified time: 2018-10-15 17:32:17
 */
import React, { Component } from 'react';

import { Menu, Icon } from 'antd';

const SubMenu = Menu.SubMenu;

import DragMenuItem from './DragMenuItem';

export default class DraggableMenu extends Component {
	constructor (props) {
		super(props);

		this.state = {
			menuDatas: [],
			openKeys: [],
			shellStyleDatas: [],
			currentShellStyle: {},
		};
	}

    componentDidMount = () => {
    	this.loadMenuDatas();

    	this.loadShellStyleDatas();
    }

	loadMenuDatas = () => {
		const { type = 'component' } = this.props;

		fetch(type == 'component' ? '../../../mock/menuDatas.json' : '../../../mock/departmentDatas.json')
			.then(result => result.json())
			.then(result => {
				const { data } = result;
				const menuDatas = this.handleMenuGroup(data);

				this.setState({ menuDatas, openKeys: menuDatas.length != 0 ? [menuDatas[0].group] : [] });
			});
	}

	loadShellStyleDatas = () => {
		fetch('../../../mock/shellStyleDatas.json')
			.then(result => result.json())
			.then(result => {
				if(result.data.length != 0)
					this.setState({ shellStyleDatas: result.data, currentShellStyle: result.data[0].style });
			});
	}

	handleMenuGroup = dataSource => {
		let result = [], groupDatas = [], tempGroupDatas = [];

		// 分出有哪几组及组别名称
		for(let item of dataSource) {
			const { group, groupName } = item;

			tempGroupDatas.push(JSON.stringify({ group, groupName }));
		}
		tempGroupDatas = Array.from(new Set(tempGroupDatas));

		for(let item of tempGroupDatas) {
			groupDatas.push(JSON.parse(item));
		}

		// 凑菜单数据结构，[{ groupName: '', children: [{ text: '', url: '' }] }]
		for(let item of groupDatas) {
			const { group, groupName } = item;
			let childResult = { groupName, group, children: [] };

			for(let jtem of dataSource) {
				const { group: groupChild } = jtem;
				const { children } = childResult;

				if(groupChild == group) {
					children.push(jtem);
				}
			}
			result.push(childResult);
		}

		return result;
	}

	handleOnOpenChange = openKeys => this.setState({ openKeys });

	handleShellStyle = currentShellStyle => this.setState({ currentShellStyle });

    render = () => {
    	const { menuDatas, openKeys, shellStyleDatas, currentShellStyle } = this.state;
    	const { type = 'component' } = this.props;

    	const menuProps = {
    		// style: { width: 256 },
    		openKeys,
    		mode: 'inline',
    		onOpenChange: this.handleOnOpenChange
    	};

    	const styleSubTitle = (
    		<span>
    			<Icon type='retweet' theme='outlined' />
    			<span>样式选择</span>
    		</span>
    	);

    	const styleSubMenu = (
    		<Menu>
    			<SubMenu title={ styleSubTitle }>
    				{
    					shellStyleDatas.map((item, i) => {
    						const { thumbnailColor, text, style } = item;

    						return (
    							<Menu.Item key={ `shellStyle${ i }` } onClick={ () => this.handleShellStyle(style) }>
    								<div style={{ width: 10, height: 10, marginRight: 10, background: thumbnailColor, display: 'inline-block' }} />
    								<span style={{ userSelect: 'none' }}>{ text }</span>
    							</Menu.Item>
    						);
    					})
    				}
    			</SubMenu>
    		</Menu>
    	);

    	return (
    		<div className='DraggableMenu'>
    			{ type == 'component' && styleSubMenu }

    			<Menu { ...menuProps }>
    				{
    					menuDatas.map(item => {
    						const { groupName, children, group } = item;

    						const subMenuTitle = (
    							<span>
    								<Icon type='laptop' theme='outlined' />
    								<span>{ groupName }</span>
    							</span>
    						);

    						return (
    							<SubMenu key={ group } title={ subMenuTitle }>
    								{ children.map(jtem => <DragMenuItem key={ `dragMenuItem${ jtem.key }` } item={ Object.assign({}, jtem, { style: currentShellStyle }) } />) }
    							</SubMenu>
    						);
    					})
    				}
    			</Menu>
    		</div>
    	);
    }
}