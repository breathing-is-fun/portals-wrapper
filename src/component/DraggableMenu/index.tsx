import React, { Component } from 'react';
import { Menu, Icon } from 'antd';
const SubMenu = Menu.SubMenu;
import DragMenuItem from './DragMenuItem';

export interface DraggableMenuProps {
  selectedKeys: Array<string>;
  menuDatas: Array<any>;
  openKeys: Array<string>;
  onClick: (group: string, keyPath: Array<string>, id: string | number) => void;
  onOpenChange: (openKeys: Array<string>) => void;
  onSave?: () => void;
  shellStyleDatas: Array<any>;
  type: 'component' | 'module' | string;
}
export interface DraggableMenuState {
  currentShellStyle: any;
}

export default class DraggableMenu extends Component<
  DraggableMenuProps,
  DraggableMenuState
> {
  private menuSelectPrefix: string;
  static defaultProps = {
    selectedKeys: [],
    menuDatas: [],
    openKeys: [],
    shellStyleDatas: [],
    type: 'component',
  };

  constructor(props: DraggableMenuProps) {
    super(props);
    this.state = {
      currentShellStyle: {},
    };
    this.menuSelectPrefix = 'dragMenuItem';
  }

  handleMenuClick = ({ item, keyPath }: any) => {
    const { onClick } = this.props;

    const { item: propsItem = {} } = item.props;

    onClick &&
      onClick(
        propsItem ? propsItem.group : null,
        [keyPath[keyPath.length - 1] + propsItem.id],
        propsItem.id,
      );
  };

  handleShellStyle = (currentShellStyle: any) => {
    this.setState({ currentShellStyle });
  };

  generateSubMenu = (children: any, currentShellStyle: any) =>
    children.map((item: any, j: number) => {
      const { group, id } = item;
      const style = Object.assign({}, item, { style: currentShellStyle });

      return (
        <DragMenuItem
          key={`${this.menuSelectPrefix}${group}${id}`}
          item={style}
        />
      );
    });

  render = () => {
    const {
      type,
      // shellStyleDatas,
      menuDatas,
      openKeys,
      selectedKeys,
      onOpenChange,
      onSave,
    } = this.props;
    const { currentShellStyle } = this.state;

    const hrefToDisplay = (
      <Menu.Item key="back" onClick={() => (location.hash = '/display')}>
        <Icon type="arrow-left" theme="outlined" />
        <span>首页</span>
      </Menu.Item>
    );

    const styleSubMenu = (
      <Menu.Item key="back" onClick={() => onSave && onSave()}>
        <Icon type="arrow-left" theme="outlined" />
        <span>保存并返回</span>
      </Menu.Item>
    );

    return (
      <div className="DraggableMenu">
        <Menu
          openKeys={openKeys}
          selectedKeys={[this.menuSelectPrefix + selectedKeys[0]]}
          mode="inline"
          onOpenChange={onOpenChange}
          onClick={this.handleMenuClick}
        >
          {type == 'module' && hrefToDisplay}

          {type == 'component' && styleSubMenu}

          {menuDatas.map(item => {
            const { groupName, children = [], group } = item;

            const subMenuTitle = (
              <span>
                <Icon type="laptop" theme="outlined" />
                <span>{groupName}</span>
              </span>
            );

            return (
              <SubMenu key={group} title={subMenuTitle}>
                {this.generateSubMenu(children, currentShellStyle)}
              </SubMenu>
            );
          })}
        </Menu>
      </div>
    );
  };
}
