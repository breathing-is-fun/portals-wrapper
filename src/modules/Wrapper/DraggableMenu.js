/*
 * @Author: zy9@github.com/zy410419243
 * @Date: 2018-09-28 17:29:59
 * @Last Modified by: zy9
 * @Last Modified time: 2018-09-29 14:24:30
 */
import React, { Component } from 'react';

import { Menu, Icon } from 'antd';

const SubMenu = Menu.SubMenu;

import DragMenuItem from './DragMenuItem';

export default class DraggableMenu extends Component {
	constructor (props) {
		super(props);

		this.state = {
			menuDatas: [],
			openKeys: [],
		};
	}

    componentDidMount = () => {
    	this.loadMenuDatas();
    }

	loadMenuDatas = () => {
		fetch('../../../mock/menuDatas.json')
			.then(result => result.json())
			.then(result => {
				const { data } = result;
				const menuDatas = this.handleMenuGroup(data);

				this.setState({ menuDatas, openKeys: menuDatas.length != 0 ? [menuDatas[0].group] : [] });
			});
	}

	handleMenuGroup = dataSource => {
		let result = [], groupDatas = [], tempGroupDatas = [];

		// 分出有哪几组及组别名称
		for(let item of dataSource) {
			const { group, groupName } = item;

			tempGroupDatas.push(JSON.stringify({ group, groupName }));
		}
		tempGroupDatas = Array.from(new Set(tempGroupDatas));

		for(let item of tempGroupDatas) {
			groupDatas.push(JSON.parse(item));
		}

		// 凑菜单数据结构，[{ groupName: '', children: [{ text: '', url: '' }] }]
		for(let item of groupDatas) {
			const { group, groupName } = item;

			let childResult = { groupName, group, children: [] };

			for(let jtem of dataSource) {
				const { group: groupChild } = jtem;
				const { children } = childResult;

				if(groupChild == group) {
					children.push(jtem);
				}
			}
			result.push(childResult);
		}

		return result;
	}

	handleOnOpenChange = openKeys => this.setState({ openKeys });

    render = () => {
    	const { menuDatas, openKeys } = this.state;
    	const menuProps = {
    		// onClick: this.handleMenuClick,
    		style: { width: 256 },
    		openKeys,
    		mode: 'inline',
    		onOpenChange: this.handleOnOpenChange
    	};

    	return (
    		<div className='Menu'>
    			<Menu { ...menuProps } ref={ ref => this.menuIns = ref }>
    				{
    					menuDatas.map(item => {
    						const { groupName, children, group } = item;

    						return (
    							<SubMenu key={ group } title={ <span><Icon type='laptop' theme='outlined' /><span>{ groupName }</span></span> }>
    								{
    									children.map(jtem => {
    										// const { key } = jtem;
    										const { draggable = true, text, key, url } = jtem;

    										return <DragMenuItem key={ `dragMenuItem${ key }` } item={ jtem } />;
    										// return <Menu.Item key={ `menuItem${ key }` } draggable={ draggable } onDragStart={ e => console.log(e) }>{ text }</Menu.Item>;
    									})
    								}
    							</SubMenu>
    						);
    					})
    				}
    			</Menu>
    		</div>
    	);
    }
}