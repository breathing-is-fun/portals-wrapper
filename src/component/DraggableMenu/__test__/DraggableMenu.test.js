/*
 * @Author: zy9@github.com/zy410419243
 * @Date: 2018-10-24 20:40:30
 * @Last Modified by: zy9
 * @Last Modified time: 2018-10-25 16:00:44
 */
import React from 'react';
import { render, mount, shallow } from 'enzyme';
import DraggableMenu from '..';

describe('DraggableMenu', () => {
	const props = {
		selectedKeys: ['all'],
		menuDatas: [{
			groupName: '全部',
			group: 'all',
			children: [
				{
					group: 'all',
					groupName: '全部',
					key: 'all',
					order: 0,
					text: '全部',
				}
			]
		}],
		openKeys: ['all'],
		onOpenChange: jest.fn(),
		onClick: jest.fn(),
	};

	it('module edit render correctly', () => {
		const wrapper = shallow(<DraggableMenu { ...Object.assign({}, props, { type: 'module' }) } />);

		expect(wrapper).toMatchSnapshot();
	});

	it('component edit render correctly', () => {
		const shellStyleDatas = [
			{
				'style': {
					'border': '1px solid blue'
				},
				'thumbnailColor': 'blue',
				'text': '蓝色'
			},
		];

		const wrapper = shallow(<DraggableMenu { ...Object.assign({}, props, { shellStyleDatas }) } />);

		expect(wrapper).toMatchSnapshot();
	});

	it('when clicked, click event should be called', () => {
		const wrapper = mount(<DraggableMenu { ...Object.assign({}, props, { type: 'module' }) } />);

		wrapper.find('.ant-menu-submenu-title').at(0).simulate('click');
		expect(props.onOpenChange.mock.calls.length).toBe(1);

		wrapper.find('.ant-menu-item').at(0).simulate('click');
		expect(props.onClick.mock.calls.length).toBe(1);
	});
});