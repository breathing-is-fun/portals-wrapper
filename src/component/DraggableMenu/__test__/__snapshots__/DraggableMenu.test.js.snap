// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DraggableMenu component edit render correctly 1`] = `
<div
  className="DraggableMenu"
>
  <Menu
    className=""
    focusable={false}
    mode="inline"
    onClick={[Function]}
    onOpenChange={[MockFunction]}
    openKeys={
      Array [
        "all",
      ]
    }
    prefixCls="ant-menu"
    selectedKeys={
      Array [
        "dragMenuItemall",
      ]
    }
    theme="light"
  >
    <MenuItem
      key="back"
      onClick={[Function]}
    >
      <Icon
        theme="outlined"
        type="arrow-left"
      />
      <span>
        保存并返回
      </span>
    </MenuItem>
    <SubMenu
      key="all"
      title={
        <span>
          <Icon
            theme="outlined"
            type="laptop"
          />
          <span>
            全部
          </span>
        </span>
      }
    >
      <DragMenuItem
        item={
          Object {
            "group": "all",
            "groupName": "全部",
            "key": "all",
            "order": 0,
            "style": Object {},
            "text": "全部",
          }
        }
        key="dragMenuItemallundefined"
      />
    </SubMenu>
  </Menu>
</div>
`;

exports[`DraggableMenu module edit render correctly 1`] = `
<div
  className="DraggableMenu"
>
  <Menu
    className=""
    focusable={false}
    mode="inline"
    onClick={[Function]}
    onOpenChange={[MockFunction]}
    openKeys={
      Array [
        "all",
      ]
    }
    prefixCls="ant-menu"
    selectedKeys={
      Array [
        "dragMenuItemall",
      ]
    }
    theme="light"
  >
    <MenuItem
      key="back"
      onClick={[Function]}
    >
      <Icon
        theme="outlined"
        type="arrow-left"
      />
      <span>
        首页
      </span>
    </MenuItem>
    <SubMenu
      key="all"
      title={
        <span>
          <Icon
            theme="outlined"
            type="laptop"
          />
          <span>
            全部
          </span>
        </span>
      }
    >
      <DragMenuItem
        item={
          Object {
            "group": "all",
            "groupName": "全部",
            "key": "all",
            "order": 0,
            "style": Object {},
            "text": "全部",
          }
        }
        key="dragMenuItemallundefined"
      />
    </SubMenu>
  </Menu>
</div>
`;
